### ISSUE TITLE
[Feature Request] Executor API: Parametrized Logit Post Processors

### ISSUE THREAD

**Issue by DreamGenX** [NVIDIA/TensorRT-LLM#2044]
### System Info

Hello, the new executor API looks promising, but it's missing a flexible logit postprocessor API -- something that's available in vLLM and most other inference backends.

**Current implementation** of logit post-processors in the Executor API requires you to register the post-processor ahead of time in a map (string -> post-processor) and then, at request time, you have to refer to the post-processor by name. Importantly, you can't pass any additional per-request params to the post-processor.

**Example use cases:**

Any use case where the post-processor requires extra input that's only available at request-time.

- Structured output generation. Let's say you want to be able to pass a regex, grammar or json schema on per-request basis and ensure that the output conforms to it.
- Adhoc logic processing: Let's say you want to implement Min P or DRY, which are missing from TensorRT-LLM. You can do that easily with a traditional logit post-processor, but not with the executor API, since you won't be able to pass the required sampler params.
- etc.

**Possible solution A (most straightforward):**

Let `LogitPostProcessor` access arbitrary `std::vector<uint8_t> state` that we provide when building the `Request` / `SamplingParams`, there we can encode additional per-request state for the logit post processor.

**Possible solution B (more performant):**

Introduce `LogitPostProcessorFactory` that accepts `std::vector<uint8_t> state`  and returns `LogitPostProcessor` based on the `state`. This will allow us to do deserialization and other processing work only once per request, rather than once per token (which could be prohibitive in terms of performance).

---

The lack of proper logit post-processor API is the main reason I have essentially re-implemented the Executor C++ API myself (the code is not open-source, so I could not just modify it either). So if the Executor API can be made more flexible, I would love it.

---

See also here:
- https://github.com/NVIDIA/TensorRT-LLM/issues/1680#issuecomment-2137826264
- https://github.com/NVIDIA/TensorRT-LLM/issues/1771#issuecomment-2164948199

### Who can help?

@bys

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

N/A

### Expected behavior

N/A

### actual behavior

N/A

### additional notes

N/A

**Comment by user-0a** [comment]
+1, Definitely agree this is needed

**Comment by pcastonguay** [comment]
It's now possible to provide a `clientId` with each request: https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/include/tensorrt_llm/executor/executor.h#L358 and this `clientId` will be passed to the logits post-processing function as the last input parameter (`std::optional<IdType>`): https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/include/tensorrt_llm/executor/types.h#L57-L58

This would allow users to associate their own state with each request, and use that in the logits post-processor.
